migration/rails.rst

```diff
@@ -0,0 +1,221 @@
+this is to simply add a ``tenant_id`` column (or “customer\_id” column,
+etc) on every object that belongs to a tenant, and backfilling your
+existing data to have this column set correctly.
+
+When you move to a distributed multi-tenant database like Citus in the
+future, this will be a required step - but if you've done this before,
+you can simply COPY over your data, without doing any additional data
+modification.
+
+**2. Use UNIQUE constraints which include the tenant\_id**
+
+Unique constraints on values will present a problem in any distributed
+system, since it’s difficult to make sure that no two nodes accept the
+same unique value.
+
+In many cases, you can work around this problem by adding the tenant\_id
+to the constraint, effectively making objects unique inside a given
+tenant, but not guaranteeing this beyond that tenant.
+
+For example, Rails creates a primary key by default, that only includes
+the ``id`` of the record:
+
+::
+
+  Indexes:
+      "page_views_pkey" PRIMARY KEY, btree (id)
+
+You should modify that primary key to also include the tenant\_id:
+
+.. code:: sql
+
+  ALTER TABLE page_views DROP CONSTRAINT page_views_pkey;
+  ALTER TABLE page_views ADD PRIMARY KEY(id, customer_id);
+
+An exception to this rule might be an email or username column on a
+users table (unless you give each tenant their own login page), which is
+why, once you scale out, we typically recommend these to be split out
+from your distributed tables and placed as a local table on the Citus
+coordinator node.
+
+**3. Include the tenant\_id in all queries, even when you can locate an
+object using its own object\_id**
+
+The easiest way to run a typical SQL query in a distributed system
+without restrictions is to always access data that lives on a single
+node, determined by the tenant you are accessing.
+
+For this reason, once you use a distributed system like Citus, we
+recommend you always specify both the tenant\_id and an object’s own ID
+for queries, so the coordinator can locate your data quickly, and can
+route the query to a single shard - instead of going to each shard in
+the system individually and asking the shard whether it knows the given
+object\_id.
+
+Updating the Rails Application
+------------------------------
+
+You can get started by including ``gem 'activerecord-multi-tenant'``
+into your Gemfile, running ``bundle install``, and then annotating your
+ActiveRecord models like this:
+
+.. code:: ruby
+
+  class PageView < ActiveRecord::Base
+    multi_tenant :customer
+    # ...
+  end
+
+In this case ``customer`` is the tenant model, and your ``page_views``
+table needs to have a ``customer_id`` column that references the
+customer the page view belongs to.
+
+The `activerecord-multi-tenant
+<https://github.com/citusdata/activerecord-multi-tenant>`__ Gem aims to
+make it easier to implement the above data changes in a typical Rails
+application.
+
+As mentioned in the beginning, by adding ``multi_tenant :customer``
```

Comment:
Took a while to go over this because of the re-ordering. Checked with Lukas, and the re-ordering makes sense. Here instead of "in the beginning", maybe "As mentioned above"?